File: src/python/algorithms/misc/insertRemoveRandom/insert_remove_random.py
import random


class InsertRemoveRandom:
    def __init__(self):
        self.num_map = {}  # val -> index
        self.nums = []

    # O(1) average
    def insert(self, val):
        if val in self.num_map:
            return False
        self.num_map[val] = len(self.nums)
        self.nums.append(val)
        return True

    # O(1) average
    def remove(self, val):
        if val not in self.num_map:
            return False
        index = self.num_map[val]
        last = self.nums[-1]
        self.nums[index] = last
        self.num_map[last] = index
        self.nums.pop()
        del self.num_map[val]
        return True

    # O(1)
    def get_random_value(self):
        return random.choice(self.nums)


if __name__ == "__main__":
    obj = InsertRemoveRandom()

    # Test 1: insert new element → True
    r1 = obj.insert(1)
    print(f"Test 1 - insert(1) new element: {'[PASS]' if r1 == True else '[FAIL]'} result={r1} expected=True")

    # Test 2: insert duplicate → False
    r2 = obj.insert(1)
    print(f"Test 2 - insert(1) duplicate: {'[PASS]' if r2 == False else '[FAIL]'} result={r2} expected=False")

    # Test 3 & 4: insert more elements
    r3 = obj.insert(2)
    print(f"Test 3 - insert(2): {'[PASS]' if r3 == True else '[FAIL]'} result={r3} expected=True")
    r4 = obj.insert(3)
    print(f"Test 4 - insert(3): {'[PASS]' if r4 == True else '[FAIL]'} result={r4} expected=True")

    # Test 5: remove existing element → True
    r5 = obj.remove(2)
    print(f"Test 5 - remove(2) exists: {'[PASS]' if r5 == True else '[FAIL]'} result={r5} expected=True")

    # Test 6: remove already-removed element → False
    r6 = obj.remove(2)
    print(f"Test 6 - remove(2) already removed: {'[PASS]' if r6 == False else '[FAIL]'} result={r6} expected=False")

    # Test 7: remove element that never existed → False
    r7 = obj.remove(99)
    print(f"Test 7 - remove(99) never existed: {'[PASS]' if r7 == False else '[FAIL]'} result={r7} expected=False")

    # Test 8: get_random_value() returns element from current set {1, 3}
    valid8 = {1, 3}
    r8 = obj.get_random_value()
    print(f"Test 8 - get_random_value() in {{1,3}}: {'[PASS]' if r8 in valid8 else '[FAIL]'} result={r8} expected=one of {{1,3}}")

    # Test 9: get_random_value() after insert; set becomes {1, 3, 4}
    obj.insert(4)
    valid9 = {1, 3, 4}
    r9 = obj.get_random_value()
    print(f"Test 9 - get_random_value() in {{1,3,4}}: {'[PASS]' if r9 in valid9 else '[FAIL]'} result={r9} expected=one of {{1,3,4}}")

File: src/javascript/algorithms/misc/insertRemoveRandom/insertRemoveRandom.js
class InsertRemoveRandom {
    constructor() {
        this.numMap = new Map(); // val -> index
        this.nums = [];
    }

    // O(1) average
    insert(val) {
        if (this.numMap.has(val)) return false;
        this.numMap.set(val, this.nums.length);
        this.nums.push(val);
        return true;
    }

    // O(1) average
    remove(val) {
        if (!this.numMap.has(val)) return false;
        const index = this.numMap.get(val);
        const last = this.nums[this.nums.length - 1];
        this.nums[index] = last;
        this.numMap.set(last, index);
        this.nums.pop();
        this.numMap.delete(val);
        return true;
    }

    // O(1)
    getRandomValue() {
        return this.nums[Math.floor(Math.random() * this.nums.length)];
    }
}

const obj = new InsertRemoveRandom();

// Test 1: insert new element → true
const r1 = obj.insert(1);
console.log(`Test 1 - insert(1) new element: ${r1 === true ? '[PASS]' : '[FAIL]'} result=${r1} expected=true`);

// Test 2: insert duplicate → false
const r2 = obj.insert(1);
console.log(`Test 2 - insert(1) duplicate: ${r2 === false ? '[PASS]' : '[FAIL]'} result=${r2} expected=false`);

// Test 3 & 4: insert more elements
const r3 = obj.insert(2);
console.log(`Test 3 - insert(2): ${r3 === true ? '[PASS]' : '[FAIL]'} result=${r3} expected=true`);
const r4 = obj.insert(3);
console.log(`Test 4 - insert(3): ${r4 === true ? '[PASS]' : '[FAIL]'} result=${r4} expected=true`);

// Test 5: remove existing element → true
const r5 = obj.remove(2);
console.log(`Test 5 - remove(2) exists: ${r5 === true ? '[PASS]' : '[FAIL]'} result=${r5} expected=true`);

// Test 6: remove already-removed element → false
const r6 = obj.remove(2);
console.log(`Test 6 - remove(2) already removed: ${r6 === false ? '[PASS]' : '[FAIL]'} result=${r6} expected=false`);

// Test 7: remove element that never existed → false
const r7 = obj.remove(99);
console.log(`Test 7 - remove(99) never existed: ${r7 === false ? '[PASS]' : '[FAIL]'} result=${r7} expected=false`);

// Test 8: getRandomValue() returns element from current set {1, 3}
const valid8 = new Set([1, 3]);
const r8 = obj.getRandomValue();
console.log(`Test 8 - getRandomValue() in {1,3}: ${valid8.has(r8) ? '[PASS]' : '[FAIL]'} result=${r8} expected=one of {1,3}`);

// Test 9: getRandomValue() after insert; set becomes {1, 3, 4}
obj.insert(4);
const valid9 = new Set([1, 3, 4]);
const r9 = obj.getRandomValue();
console.log(`Test 9 - getRandomValue() in {1,3,4}: ${valid9.has(r9) ? '[PASS]' : '[FAIL]'} result=${r9} expected=one of {1,3,4}`);
